lib/matplotlib/gridspec.py

```diff
@@ -734,3 +734,54 @@ def subgridspec(self, nrows, ncols, **kwargs):
                 fig.add_subplot(gssub[0, i])
         """
         return GridSpecFromSubplotSpec(nrows, ncols, self, **kwargs)
+
+
+class AxesArray:
+    """
+    A container for a 1D or 2D grid arrangement of Axes.
+
+    This is used as the return type of ``subplots()``.
+
+    Formerly, ``subplots()`` returned a numpy array of Axes. For a transition period,
+    AxesArray will act like a numpy array, but all functions and properties that
+    are not listed explicitly below are deprecated.
+    """
+    def __init__(self, array):
+        self._array = array
+
+    @staticmethod
+    def _ensure_wrapped(ax_or_axs):
+        if isinstance(ax_or_axs, np.ndarray):
+            return AxesArray(ax_or_axs)
+        else:
+            return ax_or_axs
+
+    def __getitem__(self, index):
+        return self._ensure_wrapped(self._array[index])
+
+    @property
+    def ndim(self):
+        return self._array.ndim
+
+    @property
+    def shape(self):
+        return self._array.shape
+
+    @property
+    def size(self):
+        return self._array.size
+
+    @property
+    def flat(self):
+        return self._array.flat
```

Comment:
could add ravel() here too.

Reply:
I guess there is no good reason to use `flatten()` over `flat`, but inexperienced me did so in lots of places.  If we don’t want to support `flatten()`, it might be good to have a more detailed deprecation message saying what to use instead.

Reply:
Agree with adding flatten() here too.  It really doesn't cost anything to support it too and should save quite a bit of churn in old codebases.

Reply:
Thanks for the feedback. Will add `flatten()` and `ravel()` as "discouraged" but keep for maximum compatibility. I also will switch to pending deprecation. Cutting the behavior is both difficult (it's hard to overview the whole array API) and not urgent. So, let's take it really slow.

Reply:
I suspect there's also users of expressions such as `axarr.T` or `axarr.ravel("F")` (or similar...) that want to iterate in column-major order.